README.pod
=pod

=encoding UTF-8

=for stopwords nopaste pastebin

=head1 NAME

App::Nopaste - Easy access to any pastebin

=head1 VERSION

version 1.013

=head1 SYNOPSIS

    use App::Nopaste 'nopaste';

    my $url = nopaste(q{
        perl -wle 'print "Prime" if (1 x shift) !~ /^1?$|^(11+?)\1+$/' [number]
    });

    # or on the command line:
    nopaste test.pl
    => http://pastebin.com/fcba51f

=head1 DESCRIPTION

Pastebins (also known as nopaste sites) let you post text, usually code, for
public viewing. They're used a lot in IRC channels to show code that would
normally be too long to give directly in the channel (hence the name nopaste).

Each pastebin is slightly different. When one pastebin goes down (I'm looking
at you, L<http://paste.husk.org>), then you have to find a new one. And if you
usually use a script to publish text, then it's too much hassle.

This module aims to smooth out the differences between pastebins, and provides
redundancy: if one site doesn't work, it just tries a different one.

It's also modular: you only need to put on CPAN a
L<App::Nopaste::Service::Foo> module and anyone can begin using it.

=head1 INTERFACE

=head2 CLI

See the documentation in L<App::Nopaste::Command>.

=head2 C<nopaste>

    use App::Nopaste 'nopaste';

    my $url = nopaste(
        text => "Full text to paste (the only mandatory argument)",
        desc => "A short description of the paste",
        nick => "Your nickname",
        lang => "perl",
        chan => "#moose",
        private => 1, # default: 0

        # this is the default, but maybe you want to do something different
        error_handler => sub {
            my ($error, $service) = @_;
            warn "$service: $error";
        },

        warn_handler => sub {
            my ($warning, $service) = @_;
            warn "$service: $warning";
        },

        # you may specify the services to use - but you don't have to
        services => ["Shadowcat", "Gist"],
    );

    print $url if $url;

The C<nopaste> function will return the URL of the paste on
success, or C<undef> on failure.

For each failure, the C<error_handler> argument is invoked with the error
message and the service that issued it.

For each warning, the C<warn_handler> argument is invoked with the warning
message and the service that issued it.

=head1 SEE ALSO

L<WebService::NoPaste>, L<WWW::Pastebin::PastebinCom::Create>, L<Devel::REPL::Plugin::Nopaste>

L<http://perladvent.org/2011/2011-12-14.html>

=head1 COPYRIGHT AND LICENSE

Copyright 2008- Shawn M Moore.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 SUPPORT

Bugs may be submitted through L<the RT bug tracker|https://rt.cpan.org/Public/Dist/Display.html?Name=App-Nopaste>
(or L<bug-App-Nopaste@rt.cpan.org|mailto:bug-App-Nopaste@rt.cpan.org>).

=head1 AUTHOR

Shawn M Moore, <sartak@gmail.com>

=head1 CONTRIBUTORS

=for stopwords Shawn M Moore Karen Etheridge Thomas Sibley Ricardo Signes François Gannaz Justin Hunter Kevin Falcone Zoffix Znet Jesse Luehrs Dean Hamstead David Golden Rafael Kitover Sergey Romanov Zakariyya Mughal gregor herrmann vti Ævar Arnfjörð Bjarmason Сергей Романов Andrew Rodland יובל קוג'מן (Yuval Kogman) Arthur Axel 'fREW' Schmidt Dan Book Darian Anthony Patrick Bremner J. Shultz Graham Knop Jason Mills John Goulah Maximilian Gass Sebastian Paaske Tørholm Shlomi Fish Tatsuhiko Miyagawa

=over 4

=item *

Shawn M Moore <code@sartak.org>

=item *

Karen Etheridge <ether@cpan.org>

=item *

Thomas Sibley <tsibley@cpan.org>

=item *

Ricardo Signes <rjbs@cpan.org>

=item *

François Gannaz <francois.gannaz@silecs.info>

=item *

Justin Hunter <justin.d.hunter@gmail.com>

=item *

Kevin Falcone <kevin@jibsheet.com>

=item *

Zoffix Znet <cpan@zoffix.com>

=item *

Jesse Luehrs <doy@tozt.net>

=item *

Dean Hamstead <dean@fragfest.com.au>

=item *

David Golden <dagolden@cpan.org>

=item *

Rafael Kitover <rkitover@cpan.org>

=item *

Sergey Romanov <complefor@rambler.ru>

=item *

Zakariyya Mughal <zaki.mughal@gmail.com>

=item *

gregor herrmann <gregoa@debian.org>

=item *

vti <viacheslav.t@gmail.com>

=item *

Ævar Arnfjörð Bjarmason <avarab@gmail.com>

=item *

Сергей Романов <sromanov@cpan.org>

=item *

Andrew Rodland <andrew@cleverdomain.org>

=item *

יובל קוג'מן (Yuval Kogman) <nothingmuch@woobling.org>

=item *

Arthur Axel 'fREW' Schmidt <frioux@gmail.com>

=item *

Dan Book <grinnz@gmail.com>

=item *

Darian Anthony Patrick <dap@darianpatrick.com>

=item *

David Bremner <bremner@unb.ca>

=item *

David J. Shultz <djshultz@gmail.com>

=item *

Graham Knop <haarg@haarg.org>

=item *

Jason Mills <jmmills@cpan.org>

=item *

John Goulah <jgoulah@gmail.com>

=item *

Maximilian Gass <mxey@ghosthacking.net>

=item *

Sebastian Paaske Tørholm <Eckankar+github@gmail.com>

=item *

Shlomi Fish <shlomif@shlomifish.org>

=item *

Tatsuhiko Miyagawa <miyagawa@bulknews.net>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by Shawn M Moore.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Changes
Revision history for App-Nopaste

{{$NEXT}}

1.013   2019-07-23 02:00:06Z
        - amended instructions in App::Nopaste::Service::Gist regarding
          creating a gist authorization token
        - --copy now works with Control-V on X windows (PR#17, Shlomi Fish)

1.012   2018-07-08 03:56:39Z
        - add GitLab service

1.011   2017-08-26 20:31:37Z
        - fix formatting of script pod

1.010   2017-08-02 04:38:50Z
        - fix shebang in 'nopaste' executable

1.009   2017-04-13 22:08:42Z
        - fixed arguments for the Mojopaste service, to work with the default
          service provider at thorsen.pm (PR#14, Andrew Rodland)

1.008   2017-02-27 01:05:29Z
        - updated paste URLs for Debian and Ubuntu services (RT#120369, Gregor
          Herrmann)
        - deprecation warning added to the now-defunct Codepeek service

1.007   2016-09-16 02:26:28Z
        - fix handling of filename option in the Gist command (Zakariyya
          Mughal, PR#13)

1.006   2016-05-24 02:55:20Z
        - fix failing tests with other locales, environment settings

1.005   2016-05-20 02:39:52Z
        - mark App::Nopaste::Service::Codepeek as deprecated in metadata
        - the Gist command also looks in ~/.github for credentials
        - many additional tests; thanks David J. Shultz!

1.004   2015-03-05 05:18:44Z
        - fix the Gist plugin to work with github's stricter validation
          (PR #11, Tatsuhiko Miyagawa)
        - removed +x permissions on files (RT#102361)
        - mark the Codepeek service as deprecated (RT#101823)

1.003   2015-01-03 20:52:17Z
        - re-release to fix problematic $VERSION declaration (RT#101095)

1.002   2014-12-19 06:50:34Z
        - new Mojopaste plugin (thanks, Dean Hamstead!)

1.001   2014-12-13 21:21:10Z
        - fix test of installed plugins to allow for third-party plugins also
          being installed

1.000   2014-12-12 03:25:40Z
        - fix the accidental loss of the plugins() method in 0.97 (thanks,
          Dean Hamstead!)

0.99    2014-12-10 04:53:43Z
        - gist URLs now use HTTPS (thanks, Darian Anthony Patrick!)

0.98    2014-11-22 07:45:48Z
        - make install prompts a bit more intelligible

0.97    2014-11-16 19:29:13Z
        - switched the JSON backend in the Gist plugin from JSON.pm to
          JSON::MaybeXS

0.96    2014-01-29 03:24:58Z
        - added missing prereqs

0.95    2014-01-26 17:41:29Z
        - bumped the minimum version needed for the (optional) prerequisite
          WWW::Pastebin::PastebinCom::Create, to avoid errors in previous
          versions (thanks, Zoffix!)

0.94    2014-01-18 22:13:29Z
        - updated and fixed debian and ubuntu language mappings (haarg, PR#1)

0.93    2014-01-14 00:41:21Z
        - packaging converted to Dist::Zilla; handling of optional features
          fixed (see RT#91053)

0.92    2013-04-26
        Fix the language mappins for Ubuntu's pastebin (Sergey Romanov)
        Don't need to load Moose in the tests (Sergey Romanov)

0.91    2013-01-27
        Decode incoming text as utf8 (Jesse Luehrs)
        Support for GITHUB_USER and GITHUB_PASS (Rafael Kitover)
        Name the gist file `nopaste` not `nopaste.txt` for possible syntax highlighting (Rafael Kitover)
        Use decode_json not a regular expression to decode Gist response body (Shawn M Moore)
        Give a hint to perldoc App::Nopaste::Service::Gist on 401 (Shawn M Moore)

0.90    2012-10-30
        Port from Moose and MooseX::Getopt to Getopt::Long::Descriptive (frew)
            App::Nopaste is not using any interesting Moose features,
            so it doesn't make sense to pay the startup cost (even
            though it's probably dominated by network latency)

            Please report any regressions!

0.35    2012-07-17
        Don't barf when gist-ing STDIN. (Ricardo SIGNES)

0.34    2012-07-14
        Support for GitHub's v3 API [Вячеслав Тихановский (vti)]
            Thanks also to Ricardo SIGNES for the call to arms:
                http://rjbs.manxome.org/rubric/entry/1964

0.33    2011-12-15
        Support for --description in Gist (zengargoyle@gmail.com) [rt.cpan.org #73272]

0.32    2011-10-21
        Better error reporting (David Golden)
        Fix some attribution errors (Salvatore Bonaccorso) [rt.cpan.org #70434]

0.31    2011-08-30
        Better filenames for ssh service (Thomas Sibley)

0.30    2011-08-26
        Fix App::Nopaste::ssh's new filename description feature (Kevin Falcone)

0.29    2011-08-22
        Add --private to several services (François Gannaz)
        Use Class::Load instead of require "Module/Name.pm" (Shawn M Moore)
        Document --list and --private options (François Gannaz)
        Less hardcoding of argument names to better support subclassing (Jason Mills)
        New Ubuntu service (http://paste.ubuntu.com/) (gregoa@debian.org)
        SSH: Add a small suggested Apache conf that's useful for paste directories (Thomas Sibley)
        SSH: optionally use the supplied description in the paste filename (Thomas Sibley)

0.28    2011-03-22
        If LWP is producing errors, *report them* (Ricardo SIGNES)

0.27    2011-03-10
        Correct path to Pastie (Sebastian Paaske Tørholm )
        Throw an error if you specify -p and files (Shawn M Moore)
        Remove Mathbin; doy moved it to a separate dist (Shawn M Moore)

0.26    2011-02-23
        Add --open (-o) for opening the nopaste in your browser (Thomas Sibley)

0.25    2011-01-03
        Add support for $GITHUB_USER/$GITHUB_TOKEN to Gist service (Maximilian Gass)

0.24    2010-12-21
        Gist requires https (Ricardo SIGNES)

0.23    2010-11-26
        Preserve the source file's extension for ssh (Thomas Sibley)
        Use Config::GitLike instead of Config::INI::Reader (Thomas Sibley)
        Doc fixes (Thomas Sibley, Justin Hunter)
        Code style fixes (Shawn M Moore)
        Remove Rafb (Justin Hunter)
        Remove Husk (Shawn M Moore)

0.22    2010-06-10 12:58:00

lib/App/Nopaste/Service/Pastie.pm
use strict;
use warnings;
package App::Nopaste::Service::Pastie;
# ABSTRACT: Service provider for Pastie - pastie.org

our $VERSION = '1.014';

use parent 'App::Nopaste::Service';

my %languages = (
    "bash" => "13",
    "c#" => "20",
    "c/c++" => "7",
    "css" => "8",
    "diff" => "5",
    "go" => "21",
    "html (erb / rails)" => "12",
    "html / xml" => "11",
    "java" => "9",
    "javascript" => "10",
    "objective-c/c++" => "1",
    "perl" => "18",
    "php" => "15",
    "plain text" => "6",
    "python" => "16",
    "ruby" => "3",
    "ruby on rails" => "4",
    "sql" => "14",
    # hidden
    "apache" => "22",
    "clojure" => "38",
    "d" => "26",
    "erlang" => "27",
    "fortran" => "28",
    "haskell" => "29",
    "ini" => "35",
    "io" => "24",
    "lisp" => "25",
    "lua" => "23",
    "makefile" => "31",
    "nu" => "36",
    "pascal" => "17",
    "puppet" => "39",
    "scala" => "32",
    "scheme" => "33",
    "smarty" => "34",
    "tex" => "37",
    # aliases
    "sh" => "13",
    "c" => "7",
    "c++" => "7",
    "objective-C" => "1",
    "objective-C++" => "1",
    "plain" => "6",
    "raw" => "6",
    "rails" => "4",
    "html" => "11",
    "xml" => "11",
    "js" => "10",
    "make" => "31",
);

sub uri { 'http://pastie.org/' }

sub fill_form {
    my $self = shift;
    my $mech = shift;
    my %args = @_;

    my $lang_id = $languages{"plain text"};
    $lang_id = $languages{lc($args{lang})}
        if (exists $args{lang} && exists $languages{lc($args{lang})});

    $mech->submit_form(
        fields => {
            "paste[body]"          => $args{text},
            "paste[authorization]" => 'burger', # set with JS to avoid bots
            "paste[restricted]"    => $args{private},
            "paste[parser_id]"     => $lang_id,
        },
    );
}

sub return {
    my $self = shift;
    my $mech = shift;

    my $prefix ='';
    my ($id) = $mech->title =~ /\#(\d+)/;
    if (!$id) {
        ($id) = $mech->content =~ m{http://pastie.org/\d+/wrap\?key=([a-z0-9]+)};
        $prefix = 'private/';
    }
    return (0, "Could not construct paste link.") if !$id;
    return (1, "http://pastie.org/$prefix$id");
}

1;
